Implement assertion

duijf opened the pull request:
WIP PR to implement assertion.

Updated Ruud's test fixtures. The tests still fail on actually verifying the signature.

duijf committed:
fido: Add placeholder assertion module

duijf committed:
server: Generalize setSessionToRegistering

This generic @casSession@ function performs compare and swap over
session data. This ensures we don't accidentally update something where
we didn't expect it.

duijf committed:
server: Sketch the handleSignIn function

duijf committed:
fido: Add some arguments to verifyAssertion

duijf committed:
fido: Work in progress on verifying assertions

duijf committed:
readme: Mention status in the readme

duijf committed:
fido: Implement signature verification for assertions

duijf committed:
tests: Test that register and login are OK

duijf committed:
tests: Update fixtures

duijf committed:
tests: Use challenge from loginReq

arianvp commented:
Maybe we should look at how https://github.com/duo-labs/webauthn/blob/master/protocol/webauthncose/webauthncose.go implements it

arianvp commented:
This fucking standard is beyond parody! 

The Signature in the cbor value is ASN.1 encoded

signature.

   Using the function defined in [RFC8017], the signature is:
   Signature = I2OSP(R, n) | I2OSP(S, n)
   where n = ceiling(key_length / 8)

https://tools.ietf.org/html/rfc8152#section-8.1

arianvp commented:
Sorry this one. https://www.w3.org/TR/webauthn/#signature-attestation-types

The COSE one isn't asn.1 but fixed size. But the one in websauthn is for backwards compatibility reasons. Wow :/

arianvp committed:
Fix signatures

What a journey; what a ride. Assertion signatures now work!

TODO: Clean up the code a bit?

arianvp committed:
Don't allow invalid Points on Curves

This allows for nasty attacks otherwise.

Also fix some warnings about unused variables

arianvp commented:
@duijf I fixed the signature code!

I don't know what else needs to happen in this PR, so I give it back to you

duijf committed:
fido: Do not implement signature counting

duijf committed:
server: Extract all handlers to separate functions

duijf committed:
server: Clean up registerBegin

duijf committed:
server: Clean up the Text imports

duijf committed:
server: Extract relying party settings

duijf committed:
server: Implement verifyLogin

duijf committed:
js: Rename `name` to `userName`

duijf committed:
js: Add form to test authentication

duijf committed:
server: Always set cookies on `/`

This was messing up our session management and was causing the wrong
session keys to be sent to the server.

duijf committed:
server: Autoformat

duijf commented:
I managed to get this to work :tada: 

Also found a bug in our cookie settings while I was at this: we didn't set the cookie path, so we would have separate sessions per URL of our application. I changed this to set all cookies on `/`

You can test this in your browser by keeping an eye on your devtools while doing:

 - Click the "Test authentication" button and see a 401 response.
 - Registering a new credential
 - Click the "Test authentication" button and see a 200 response.
 - Click "Login" and see a 400 bad request (because you're already signed in).
 - Clear all cookies in your devtools without refreshing the page.
 - Click the "Test authentication" button and see a 401 response.
 - Click "Login" and use your Yubikey.
 - Click the "Test authentication" button and see a 200 response.

(Admittedly, this is not a convincing demo and bad UX. We can clean that up later, the most important thing is that this works!)

arianvp commented:
Cool. Just tried and it works!

How about we add a `userName` field to the  `Login` form as well; we can then look up the `userHandle` in the database and do a login even after a refresh

However we can also clean up the "UX" of the demo in a later PR.

arianvp commented on fido/Crypto/Fido2/Assertion.hs:
```diff
@@ -0,0 +1,100 @@
+data RelyingPartyConfig = RelyingPartyConfig {origin :: Fido2.Origin, rpId :: Fido2.RpId}
+
+-- | Verify that a @Fido2.PublicKeyCredential@ is valid for the given
+-- @RelyingPartyConfig@, @Challenge@, and list of @Creedential@s known to be
```
If you use ` 'RelyingPartyConfig' ` instead of `@ @ ` you get hyperlinks in the source code.

However; we should probably just do a big haddock pass when this library is ready for release, so dont bikeshed too much about it now

duijf replied:
Thanks! Cleaned up some minor things about the haddocks.

duijf commented:
> How about we add a userName field to the Login form as well; we can then look up the userHandle in the database and do a login even after a refresh

This sounds nice! I would like to wait on #15 to land before we spend time on this though

duijf committed:
fido: Clean up haddocks in assertion module

duijf commented:
Rebased and resolved conflicts. Am going to merge this
